File: app/src/main/java/io/github/gmazzo/android/livewallpaper/weather/engine/things/ThingLightning.java
package io.github.gmazzo.android.livewallpaper.weather.engine.things;

import static javax.microedition.khronos.opengles.GL10.GL_COLOR_BUFFER_BIT;
import static javax.microedition.khronos.opengles.GL10.GL_LIGHTING;
import static javax.microedition.khronos.opengles.GL10.GL_ONE;
import static javax.microedition.khronos.opengles.GL10.GL_SRC_ALPHA;

import javax.microedition.khronos.opengles.GL10;

import io.github.gmazzo.android.livewallpaper.weather.R;
import io.github.gmazzo.android.livewallpaper.weather.engine.EngineColor;
import io.github.gmazzo.android.livewallpaper.weather.engine.GlobalRand;
import io.github.gmazzo.android.livewallpaper.weather.wallpaper.Models;
import io.github.gmazzo.android.livewallpaper.weather.wallpaper.Texture;
import io.github.gmazzo.android.livewallpaper.weather.wallpaper.Textures;

public class ThingLightning extends Thing {
    private static final int MODELS[] = {
            R.raw.lightning1, R.raw.lightning2, R.raw.lightning3
    };
    private Texture lightningGlow;
    private Texture lightningCore;

    public ThingLightning(float r, float g, float b) {
        this.engineColor = new EngineColor(r, g, b, 1.0f);
    }

    @Override
    public void render(GL10 gl, Textures textures, Models models) {
        if (model == null) {
            int number = GlobalRand.intRange(1, 4);
            model = models.get(MODELS[number - 1]);
            lightningGlow = textures.get(R.raw.lightning_pieces_glow);
            lightningCore = textures.get(R.raw.lightning_pieces_core);
        }

        gl.glEnable(GL_LIGHTING);
        gl.glEnable(GL_COLOR_BUFFER_BIT);

        gl.glBlendFunc(GL_SRC_ALPHA, GL_ONE);
        gl.glPushMatrix();
        gl.glTranslatef(this.origin.getX(), this.origin.getY(), this.origin.getZ());
        gl.glScalef(this.scale.getX(), this.scale.getX(), this.scale.getX());
        gl.glRotatef(this.angles.getA(), this.angles.getR(), this.angles.getG(), this.angles.getB());
        if (this.engineColor != null) {
            gl.glColor4f(this.engineColor.getR(), this.engineColor.getG(), this.engineColor.getB(), this.engineColor.getA());
        }
        model.renderFrameMultiTexture(lightningGlow, lightningCore, 260, false);
        gl.glPopMatrix();
        gl.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
        gl.glDisable(GL_COLOR_BUFFER_BIT);
        gl.glDisable(GL_LIGHTING);
    }

    public void update(float timeDelta) {
        super.update(timeDelta);
        this.engineColor.setA(this.engineColor.getA() - 2.0f * timeDelta);
        if (this.engineColor.getA() <= 0.0f) {
            delete();
        }
    }
}

File: app/src/main/kotlin/io/github/gmazzo/android/livewallpaper/weather/engine/things/ThingLightning.kt
package io.github.gmazzo.android.livewallpaper.weather.engine.things

import io.github.gmazzo.android.livewallpaper.weather.R
import io.github.gmazzo.android.livewallpaper.weather.engine.models.Models
import io.github.gmazzo.android.livewallpaper.weather.engine.pushMatrix
import io.github.gmazzo.android.livewallpaper.weather.engine.textures.Textures
import io.github.gmazzo.android.livewallpaper.weather.engine.time.Clock
import io.github.gmazzo.android.livewallpaper.weather.engine.withColor
import io.github.gmazzo.android.livewallpaper.weather.engine.withFlags
import kotlinx.coroutines.flow.MutableStateFlow
import javax.inject.Inject
import javax.microedition.khronos.opengles.GL10.GL_ADD
import javax.microedition.khronos.opengles.GL10.GL_COLOR_BUFFER_BIT
import javax.microedition.khronos.opengles.GL10.GL_LIGHTING
import javax.microedition.khronos.opengles.GL10.GL_MODELVIEW
import javax.microedition.khronos.opengles.GL10.GL_ONE
import javax.microedition.khronos.opengles.GL10.GL_SRC_ALPHA
import javax.microedition.khronos.opengles.GL11
import kotlin.random.Random

class ThingLightning @Inject constructor(
    random: Random,
    gl: GL11,
    private val resources: Resources,
    private val clock: MutableStateFlow<Clock>,
) : Thing(
    gl,
    model = resources.models[random.nextInt(resources.models.size)],
    texture = resources.lightningPiecesCore,
) {

    override fun render() = gl.pushMatrix(GL_MODELVIEW) {
        withFlags(GL_LIGHTING, GL_COLOR_BUFFER_BIT) {
            glBlendFunc(GL_SRC_ALPHA, GL_ONE)
            glTranslatef(origin.x, origin.y, origin.z)
            glScalef(scale.x, scale.x, scale.x)

            withColor(color) {
                model.render(resources.lightningPiecesGlow, texture, GL_ADD)
            }
        }
    }

    override fun update() {
        super.update()

        color.a -= 2f * clock.value.deltaSeconds
        if (color.a <= 0f) {
            delete()
        }
    }

    class Resources @Inject constructor(
        models: Models,
        textures: Textures,
    ) {
        val lightning1 = models[R.raw.lightning1]
        val lightning2 = models[R.raw.lightning2]
        val lightning3 = models[R.raw.lightning3]
        val lightningPiecesCore = textures[R.raw.lightning_pieces_core]
        val lightningPiecesGlow = textures[R.raw.lightning_pieces_glow]

        val models = arrayOf(lightning1, lightning2, lightning3)
    }

}
